reinforcement_learning/rlclientlib/logger/eventhub_client.cc

```diff
@@ -18,13 +18,13 @@ namespace reinforcement_learning {
   eventhub_client::http_request_task::http_request_task(web::http::client::http_client& client,
     const std::string& host,
     const std::string& auth,
-    std::string&& post_data)
+    const std::vector<unsigned char>& post_data)
```

Comment:
Overall comment to all these changes. Looks like review will look much simpler if it can be splitted in few stages, where the first step is introducing new abstraction (message/buffer/...) on top of current std::string (even with just typedef). It is quite broad change, but with 0 breaking changes and very easier to validate. After that replacing it to vector<char> (or anything else) turns into quite local changes.

Reply:
True. The reason is that no one would agree to the abstraction change unless people seeing flatbuffer changes need. It's kinda chicken and egg problem.

---
In reply to: [228201579](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228201579) [](ancestors = 228201579)

Reply:
It looks like mandatory change anyways. You have already done heavy work to discriminate all std::string between sender message and other ones, but the next person who will need to introduce next change on this level have to do the same now.

Reply:
So having this abstraction is 100% good thing even if all other changes will not be accepted for some reason.

Reply:
Vector of unsigned char is not abstraction - if we decide to change it to
`class NewMessage {vector<unsigned char> data; bool someFlag;}`
it requires almost the same amount of changes as you are doing now.

I am talking about having first PR something based on:
`class i_sender
{
  using message_t = std::string;
  int send(message_t&& data, api_status* status = nullptr);
  ...
}`
This review is quite simple and will cover almost every signature change that you have in your current PR with 0 logical changes.

Remaining part will be almost completely within ranking_event.cc

Reply:
To me this abstraction is good to have, but may not a must. If we want this, I prefer we can have a follow up PR to address this.  We can discuss this offline.

I do agree if we all agree flatbuffer approach at the begining (e.g. a month back) and want to do this PR incrementally, this is the best approach.

---
In reply to: [228333588](https://github.com/VowpalWabbit/vowpal_wabbit/pull/1653#discussion_r228333588) [](ancestors = 228333588)